**Import Libraries and Files**

In [53]:
import pandas as pd
import numpy as np

In [54]:
# Load the CSV file
lists = pd.read_csv("csvs/lists_2019_8.csv",encoding="ISO-8859-1")
commits = pd.read_csv("csvs/modified_commits_2019_8.csv",encoding="ISO-8859-1")
people = pd.read_csv("csvs/people_2019_8.csv",encoding="ISO-8859-1")
# pr = pd.read_csv("csvs/pull-requests.csv",encoding="ISO-8859-1")
pr = pd.read_csv("csvs/modified_pull-requests.csv",encoding="ISO-8859-1")
alias=pd.read_csv("csvs/modified_aliases_2019_8.csv",encoding="ISO-8859-1")
repo_name=pd.read_csv("repo_name_match.csv",encoding="ISO-8859-1")

/tmp/ipykernel_261583/1026464259.py:3: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  commits = pd.read_csv("csvs/modified_commits_2019_8.csv",encoding="ISO-8859-1")


**Joining the necessary tables**

*Reading the projects*

In [56]:
projects=lists[['listid','listname','status']]

In [57]:
# Convert all values in the 'listname' column to lowercase
projects['listname'] = projects['listname'].str.lower()

/tmp/ipykernel_261583/2186385244.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  projects['listname'] = projects['listname'].str.lower()


In [187]:
projects.head(3)

listid   listname  status
0       1  Amaterasu       0
1       2  Annotator       0
2       3    BatchEE       0

*Reading the commits*

In [58]:
com=commits[['list','authorid','committerid']]

In [142]:
com.head(3)

list  authorid  committerid
0    54      2846         2846
1    54      2846         2846
2    54      2846         2846

*Reading alias*

In [59]:
devs=alias[['aliasid','name','email']]

In [60]:
# Convert all values in the 'listname' column to lowercase
devs['name'] = devs['name'].str.lower()

/tmp/ipykernel_261583/2314199456.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  devs['name'] = devs['name'].str.lower()


In [8]:
devs.head(3)

aliasid                             name  \
0    22233  andy seaborne (assigned) (jira)   
1    12937                     olivier lamy   
2    12967   olivier lamy (reopened) (jira)   

                                               email  
0                                    jira@apache.org  
1  reply+i-2355594-3a189c6b1c866515f1dd1db09f4f45...  
2                                    jira@apache.org

In [61]:
print(com.shape)
print(projects.shape)
print(devs.shape)

(3655492, 3)
(312, 3)
(47078, 3)


In [62]:
projects.rename(columns={'listid': 'project_id'}, inplace=True)
projects.rename(columns={'listname': 'project_name'}, inplace=True)
projects.rename(columns={'status': 'project_status'}, inplace=True)

/tmp/ipykernel_261583/816927746.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  projects.rename(columns={'listid': 'project_id'}, inplace=True)
/tmp/ipykernel_261583/816927746.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  projects.rename(columns={'listname': 'project_name'}, inplace=True)
/tmp/ipykernel_261583/816927746.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  projects.rename(columns={'status': 'project_status'}, i

In [64]:
projects.columns

Index(['project_id', 'project_name', 'project_status'], dtype='object')

In [63]:
com.rename(columns={'list': 'project_id'}, inplace=True)
com.rename(columns={'authorid': 'alias_id'}, inplace=True)

/tmp/ipykernel_261583/3586036795.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  com.rename(columns={'list': 'project_id'}, inplace=True)
/tmp/ipykernel_261583/3586036795.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  com.rename(columns={'authorid': 'alias_id'}, inplace=True)


In [65]:
com.columns

Index(['project_id', 'alias_id', 'committerid'], dtype='object')

In [66]:
devs.rename(columns={'aliasid': 'alias_id'}, inplace=True)
devs.rename(columns={'name': 'dev_name'}, inplace=True)

/tmp/ipykernel_261583/74888050.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  devs.rename(columns={'aliasid': 'alias_id'}, inplace=True)
/tmp/ipykernel_261583/74888050.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  devs.rename(columns={'name': 'dev_name'}, inplace=True)


In [15]:
devs.columns

Index(['alias_id', 'dev_name', 'email'], dtype='object')

**Right Outer join projects and commits table**

In [67]:
# Right Outer Join
projects_commits = projects.merge(com, on='project_id', how='right')

In [68]:
projects_commits.head(10)

project_id project_name  project_status  alias_id  committerid
0          54      airflow               1      2846         2846
1          54      airflow               1      2846         2846
2          54      airflow               1      2846         2846
3          54      airflow               1      2846         2846
4          54      airflow               1      2846         2846
5          54      airflow               1      2846         2846
6          54      airflow               1      2846         2846
7          54      airflow               1      2846         2846
8          54      airflow               1      2846         2846
9          54      airflow               1      2846         2846

In [69]:
projects_commits.shape

(3655492, 5)

**This means we got 3655492 commits from 263 projects**

In [70]:
unique_values = projects_commits['project_id'].unique()
print(unique_values)
print(unique_values.size)

[ 54 256 132  55 129 128 276 254   1  56   2  58  59  49 277  65 130  61
  62  63  64 257   3  70  67  68  69 259 260  81  71  66  72  73  74  75
  76  77  78  79  80 179  83 261  82 263  92 264  84   5  85  86  87  88
   6  89  90  91  13  93  94  95  96   8  10  97 267  11  98  99 100  12
 106 113  14 107 108 109 110 112 268 101 102 103 104 105  15 121  57 116
 269 270 176 117 177 178 181 118 183 184 185 186 188 291  52 190 119  29
 292 191 192 195  50 196 197 135 136 120 137 139  51 279 142 280 143 145
  21 283 154 156 157 159 160 161 163 285 164  24  25  26 168 169 286 287
 170 171 172 173 289 115 122 131 223 199 295 200 296 201 123 203 271  33
 297 124 204 205 207 208 209 210  37 298 211 272  16 273  17 212 213 214
 215 216 217 218 127 219 301 253  53  39 225 302  41 229 230 231 232  43
 235 303 238 240 241 242 243 244 306 246 248 250 308 310 252 312  18  20
 153 288  30  31  45  48 134 138 141 281 144 146 155  23 167 175  27 182
 290 189 187  28 194 293 294 206 202  32  34  35 29

In [71]:
# Right Outer Join on people_id
merged = devs.merge(projects_commits, on='alias_id', how='right')

In [72]:
merged.head(10)

alias_id   dev_name                 email  project_id project_name  \
0      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
1      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
2      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
3      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
4      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
5      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
6      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
7      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
8      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   
9      2846  kaxilnaik  kaxilnaik@apache.org          54      airflow   

   project_status  committerid  
0               1         2846  
1               1         2846  
2               1         2846  
3               1         2846  
4               1         2846  
5               1         2846  
6               1         2846  
7               1         2846  
8               1         2846  
9               1         2846

In [73]:
merged.shape

(3655492, 7)

**So we finally got all the commit details with relevant projects and committers**

In [22]:
# Check if a column contains NaN
has_nan = merged['dev_name'].isna().any()

print(has_nan)  # We have some people with no names

True


In [210]:
# Filter rows which has atleast one nan value
rows_with_nan = merged[merged['dev_name'].isna()]

# print(rows_with_nan)

# Count occurrences of each unique value in a column
unique_value_counts = rows_with_nan['alias_id'].value_counts()

print(unique_value_counts)

alias_id
12991    8
Name: count, dtype: int64


**We determine Core Developers here**

In [74]:
df = merged

# Count total commits per person per project
commit_counts = df.groupby(['project_id', 'alias_id']).size().reset_index(name='commit_count')

# Count total commits per project
total_commits_per_project = df.groupby('project_id').size().reset_index(name='total_commits')

# Count unique people per project
num_people_per_project = df.groupby('project_id')['alias_id'].nunique().reset_index(name='num_people')

# Merge all the information
result_df = commit_counts.merge(total_commits_per_project, on='project_id').merge(num_people_per_project, on='project_id')

# Calculate average commits per person in the project
result_df['average_commits'] = result_df['total_commits'] / result_df['num_people']

# Calculate percentage of total commits each person made
result_df['commit_percentage'] = (result_df['commit_count'] / result_df['total_commits']) * 100

# Sort by project_id first, then by commit_percentage in descending order
result_df = result_df.sort_values(by=['project_id', 'commit_percentage'], ascending=[True, False])

# Calculate cumulative sum of commit percentage within each project
result_df['cumulative_commit_percentage'] = result_df.groupby('project_id')['commit_percentage'].cumsum()

# Identify Core Developers
def identify_core_devs(group):
    core_devs = []
    cumulative_percentage = 0
    
    for index, row in group.iterrows():
        if row['commit_count'] < row['average_commits']:  # Stop if below average commits
            break
        cumulative_percentage += row['commit_percentage']
        core_devs.append(index)
        if cumulative_percentage >= 90:  # Stop if cumulative commits reach 90%+
            break
    
    group['Core_Developer'] = 'No'
    group.loc[core_devs, 'Core_Developer'] = 'Yes'
    group['Total_Core_Developers'] = group['Core_Developer'].eq('Yes').sum()
    return group

# Apply function per project
result_df = result_df.groupby('project_id', group_keys=False).apply(identify_core_devs)

# Display result
print(result_df)

      project_id  alias_id  commit_count  total_commits  num_people  \
0              1      4097            39             39           1   
5              2      4427           110            161           6   
2              2      4423            22            161           6   
4              2      4425            21            161           6   
6              2      4428             4            161           6   
...          ...       ...           ...            ...         ...   
7815         311     45916            35          13625           7   
7817         311     45919            12          13625           7   
7818         312     21969            52            102           3   
7819         312     47081            30            102           3   
7820         312     47083            20            102           3   

      average_commits  commit_percentage  cumulative_commit_percentage  \
0           39.000000         100.000000                    100.000000   

/tmp/ipykernel_261583/2453455463.py:46: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  result_df = result_df.groupby('project_id', group_keys=False).apply(identify_core_devs)


In [75]:
result_df.head(25)

project_id  alias_id  commit_count  total_commits  num_people  \
0            1      4097            39             39           1   
5            2      4427           110            161           6   
2            2      4423            22            161           6   
4            2      4425            21            161           6   
6            2      4428             4            161           6   
3            2      4424             3            161           6   
1            2      4422             1            161           6   
7            3      4099          3037           8287          13   
10           3      5924          2603           8287          13   
19           3      5933          1242           8287          13   
12           3      5926           476           8287          13   
11           3      5925           363           8287          13   
15           3      5929           217           8287          13   
16           3      5930           174           8287          13   
17           3      5931            79           8287          13   
14           3      5928            39           8287          13   
13           3      5927            31           8287          13   
8            3      4107            21           8287          13   
9            3      5892             3           8287          13   
18           3      5932             2           8287          13   
21           5     11519          1531           3520          15   
27           5     11525           606           3520          15   
23           5     11521           478           3520          15   
31           5     11530           343           3520          15   
22           5     11520           162           3520          15   

    average_commits  commit_percentage  cumulative_commit_percentage  \
0         39.000000         100.000000                    100.000000   
5         26.833333          68.322981                     68.322981   
2         26.833333          13.664596                     81.987578   
4         26.833333          13.043478                     95.031056   
6         26.833333           2.484472                     97.515528   
3         26.833333           1.863354                     99.378882   
1         26.833333           0.621118                    100.000000   
7        637.461538          36.647762                     36.647762   
10       637.461538          31.410643                     68.058405   
19       637.461538          14.987330                     83.045734   
12       637.461538           5.743936                     88.789671   
11       637.461538           4.380355                     93.170025   
15       637.461538           2.618559                     95.788585   
16       637.461538           2.099674                     97.888259   
17       637.461538           0.953300                     98.841559   
14       637.461538           0.470617                     99.312176   
13       637.461538           0.374080                     99.686256   
8        637.461538           0.253409                     99.939665   
9        637.461538           0.036201                     99.975866   
18       637.461538           0.024134                    100.000000   
21       234.666667          43.494318                     43.494318   
27       234.666667          17.215909                     60.710227   
23       234.666667          13.579545                     74.289773   
31       234.666667           9.744318                     84.034091   
22       234.666667           4.602273                     88.636364   

   Core_Developer  Total_Core_Developers  
0             Yes                      1  
5             Yes                      1  
2              No                      1  
4              No                      1  
6              No                      1  
3              No                      1  
1              No                  

**Saving the core developers per project**

In [76]:
result_df.to_csv("project_core_dev.csv",index=False)

**Connecting Core Developer with Pull Request**

In [77]:
print(pr.head(10))

       repo  number                  user user_type   state  \
0    abdera       4               dargooo      User  closed   
1    abdera       3              YYTVicky      User    open   
2    abdera       2  nicolasyanncouturier      User    open   
3    abdera       1  nicolasyanncouturier      User    open   
4  accumulo    4027              dlmarion      User  closed   
5  accumulo    4026              dlmarion      User  closed   
6  accumulo    4025              dlmarion      User  closed   
7  accumulo    4024              ctubbsii      User  closed   
8  accumulo    4023              dlmarion      User  closed   
9  accumulo    4022              dlmarion      User  closed   

             created_at            updated_at             closed_at  \
0  2020-11-26T21:21:33Z  2023-05-21T04:37:38Z  2023-05-21T04:37:38Z   
1  2020-03-01T00:37:56Z  2020-04-30T02:01:45Z                   NaN   
2  2013-10-18T13:32:37Z  2014-07-16T01:44:57Z                   NaN   
3  2013-10-14T12:23:03

In [27]:
pr.head(10)

repo  number                  user user_type   state  \
0    abdera       4               dargooo      User  closed   
1    abdera       3              YYTVicky      User    open   
2    abdera       2  nicolasyanncouturier      User    open   
3    abdera       1  nicolasyanncouturier      User    open   
4  accumulo    4027              dlmarion      User  closed   
5  accumulo    4026              dlmarion      User  closed   
6  accumulo    4025              dlmarion      User  closed   
7  accumulo    4024              ctubbsii      User  closed   
8  accumulo    4023              dlmarion      User  closed   
9  accumulo    4022              dlmarion      User  closed   

             created_at            updated_at             closed_at  \
0  2020-11-26T21:21:33Z  2023-05-21T04:37:38Z  2023-05-21T04:37:38Z   
1  2020-03-01T00:37:56Z  2020-04-30T02:01:45Z                   NaN   
2  2013-10-18T13:32:37Z  2014-07-16T01:44:57Z                   NaN   
3  2013-10-14T12:23:03Z  2014-07-13T19:17:32Z                   NaN   
4  2023-12-06T14:20:37Z  2024-07-12T00:15:54Z  2023-12-06T14:34:06Z   
5  2023-12-06T14:02:03Z  2024-07-12T00:15:53Z  2023-12-06T14:17:41Z   
6  2023-12-06T13:02:38Z  2024-07-12T00:15:53Z  2023-12-06T14:18:21Z   
7  2023-12-05T23:12:06Z  2024-07-12T01:30:39Z  2023-12-06T19:46:39Z   
8  2023-12-05T21:54:27Z  2024-07-12T00:15:52Z  2023-12-06T12:27:43Z   
9  2023-12-05T20:03:13Z  2024-07-12T00:15:52Z  2023-12-08T14:04:26Z   

              merged_at                          merge_commit_sha  \
0                   NaN  85551716e4900cb514c5913086cedb869edb5ecf   
1                   NaN  10b439abf352089d813cf882e9d6291fb58bd363   
2                   NaN  e973e7069a3e4be285b2b934d57adf32caa0a9ef   
3                   NaN  975791491d738b7d0467bd1975e799857a1c43ed   
4  2023-12-06T14:34:06Z  29c307fca48117de1433754ead350e031c4f4153   
5  2023-12-06T14:17:41Z  7ee79949b88a01dfd2fd133d267197ad7138d807   
6  2023-12-06T14:18:21Z  c340c7aaec4f202a76a51f28633ebc931426a56e   
7  2023-12-06T19:46:38Z  a3444f7fda056bed6e36f8b3bc6252abbe1807cb   
8  2023-12-06T12:27:43Z  8f88fbe17cf86d6500ffe4d5c1477d0ecc2008cd   
9  2023-12-08T14:04:26Z  ce7c8c8f2731e0f7aae679b735050ca63a7be8d1   

                                           issue_url  \
0  https://api.github.com/repos/apache/abdera/iss...   
1  https://api.github.com/repos/apache/abdera/iss...   
2  https://api.github.com/repos/apache/abdera/iss...   
3  https://api.github.com/repos/apache/abdera/iss...   
4  https://api.github.com/repos/apache/accumulo/i...   
5  https://api.github.com/repos/apache/accumulo/i...   
6  https://api.github.com/repos/apache/accumulo/i...   
7  https://api.github.com/repos/apache/accumulo/i...   
8  https://api.github.com/repos/apache/accumulo/i...   
9  https://api.github.com/repos/apache/accumulo/i...   

                                                body  project_id  
0                                                NaN        49.0  
1                                                NaN        49.0  
2  Used the same system in BaseSerializer.writeEx...        49.0  
3  The test would most likely need to be split in...        49.0  
4                                                NaN        50.0  
5                                                NaN        50.0  
6  The exception handling for the BadLocationStat...        50.0  
7  Rewrite MemoryConsumingIterator's method to co...        50.0  
8  Calling getServerContext() ends up down a path...        50.0  
9                                        Fixes #4021        50.0

In [78]:
# Convert all values in the 'listname' column to lowercase
pr['repo'] = pr['repo'].str.lower()

**Among 252 projects in PR Lists, 62 have "incubator-" written in front of them which is causing string mismatch with project names in Lists
Again, 15 among the 62 have "retired-" written after "incubator-". We need to remove them both.**

In [127]:
# Remove "incubator-" from the start of strings
pr["repo"] = pr["repo"].str.replace(r"^incubator-", "", regex=True)
pr["repo"] = pr["repo"].str.replace(r"^retired-", "", regex=True)

In [ ]:
# Convert all values in the 'listname' column to lowercase
pr['repo'] = pr['repo'].str.lower()

# We search every repo names in pr to repo_name and trace out the project_id and then add it to pr
pr['project_id'] = pr['repo'].apply(
    lambda repo: repo_name.loc[repo_name['matched_repo'] == repo, 'listid'].values[0] if not repo_name.loc[repo_name['matched_repo'] == repo, 'listid'].empty else np.nan
)

In [ ]:
pr.head(10)

In [ ]:
# Save the file if necessary for further processing
# pr.to_csv("csvs/modified_pull-requests.csv",index=False)

In [66]:
unique_values = pr['state'].unique()
print(unique_values)
print(unique_values.size)

['closed' 'open']
2


In [36]:
unique_values = pr['number'].unique()
print(unique_values)
print(unique_values.size)

[    4     3     2 ... 48961 48960 48959]
49853


In [37]:
unique_values = pr['user'].unique()
print(unique_values)
print(unique_values.size)

['dargooo' 'YYTVicky' 'nicolasyanncouturier' ... 'bulicekj' 'zpf-github'
 'hyangtack']
39241


In [28]:
unique_values = pr['repo'].unique()
print(unique_values)
print(unique_values.size)

['abdera' 'accumulo' 'ace' 'activemq' 'activemq-apollo' 'airavata'
 'airflow' 'allura' 'ambari' 'ant-easyant-core' 'ant-ivy' 'any23'
 'apex-core' 'aries' 'asterixdb' 'atlas' 'aurora' 'beam' nan 'bigtop'
 'bloodhound' 'brooklyn' 'buildr' 'bval' 'calcite' 'carbondata'
 'cassandra' 'cayenne' 'celix' 'chukwa' 'clerezza' 'climate' 'cloudstack'
 'commons-ognl' 'cordova' 'couchdb' 'creadur-rat' 'crunch' 'ctakes'
 'curator' 'cxf' 'datafu' 'db-jdo' 'deltacloud' 'deltaspike' 'derby'
 'directmemory' 'directory-server' 'distributedlog' 'drill' 'dubbo'
 'eagle' 'empire-db' 'esme' 'falcon' 'felix' 'fineract' 'flex-sdk' 'flink'
 'flume' 'fluo' 'freemarker' 'ftpserver' 'geode' 'geronimo'
 'geronimo-yoko' 'giraph' 'gora' 'griffin' 'groovy' 'guacamole-client'
 'hama' 'hawq' 'hcatalog' 'helix' 'ignite' 'impala' 'amaterasu'
 'annotator' 'ariatosca' 'batchee' 'brpc' 'crail' 'daffodil'
 'datasketches' 'dlab' 'dolphinscheduler' 'doris' 'druid' 'echarts'
 'edgent' 'flagon' 'gobblin' 'heron' 'hivemall' 'hudi' 

In [107]:
a=projects[['project_id','project_status']]
b=result_df[['project_id','alias_id','Core_Developer']]
c=pr[['repo','user','created_at','updated_at','closed_at','merged_at','state','project_id']]
d=devs

In [108]:
print(a.shape)
print(b.shape)
print(c.shape)
print(d.shape)

(312, 2)
(7821, 3)
(569448, 8)
(47078, 3)


In [70]:
d.columns

Index(['alias_id', 'dev_name', 'email'], dtype='object')

In [56]:
a.columns

Index(['project_id', 'project_name', 'project_status'], dtype='object')

In [50]:
b.shape

(7821, 3)

In [51]:
unique_values = b['Core_Developer'].unique()
print(unique_values)
print(unique_values.size)

['Yes' 'No']
2


In [83]:
b = b.drop_duplicates(keep='first')

In [84]:
b.shape

(7821, 3)

In [85]:
# Check if there are any duplicate rows
duplicates_exist = b.duplicated().any()

print(duplicates_exist)

False


In [86]:
# Filter rows where project_id is '54' and select Core_developer column
core_dev_values = b.loc[b['project_id'] == 54, 'Core_Developer']

# Print the results
print("Core_developer values for project_id:")
print(core_dev_values.to_list())  # Converts to a list for better readability

Core_developer values for project_id:
['Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No

In [87]:
b.head(10)

project_id  alias_id Core_Developer
0            1      4097            Yes
5            2      4427            Yes
2            2      4423             No
4            2      4425             No
6            2      4428             No
3            2      4424             No
1            2      4422             No
7            3      4099            Yes
10           3      5924            Yes
19           3      5933            Yes

In [109]:
c.rename(columns={'user': 'dev_name'}, inplace=True)
c.rename(columns={'state': 'pr_status'}, inplace=True)
c.rename(columns={'repo': 'project_name'}, inplace=True)
d.rename(columns={'personid': 'alias_id'}, inplace=True)

/tmp/ipykernel_261583/2494848850.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c.rename(columns={'user': 'dev_name'}, inplace=True)
/tmp/ipykernel_261583/2494848850.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c.rename(columns={'state': 'pr_status'}, inplace=True)
/tmp/ipykernel_261583/2494848850.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  c.rename(columns={'repo': 'project_name'}, inplace=True)
/tmp/ipykernel_2615

In [110]:
c.head(10)

project_name              dev_name            created_at  \
0       abdera               dargooo  2020-11-26T21:21:33Z   
1       abdera              YYTVicky  2020-03-01T00:37:56Z   
2       abdera  nicolasyanncouturier  2013-10-18T13:32:37Z   
3       abdera  nicolasyanncouturier  2013-10-14T12:23:03Z   
4     accumulo              dlmarion  2023-12-06T14:20:37Z   
5     accumulo              dlmarion  2023-12-06T14:02:03Z   
6     accumulo              dlmarion  2023-12-06T13:02:38Z   
7     accumulo              ctubbsii  2023-12-05T23:12:06Z   
8     accumulo              dlmarion  2023-12-05T21:54:27Z   
9     accumulo              dlmarion  2023-12-05T20:03:13Z   

             updated_at             closed_at             merged_at pr_status  \
0  2023-05-21T04:37:38Z  2023-05-21T04:37:38Z                   NaN    closed   
1  2020-04-30T02:01:45Z                   NaN                   NaN      open   
2  2014-07-16T01:44:57Z                   NaN                   NaN      open   
3  2014-07-13T19:17:32Z                   NaN                   NaN      open   
4  2024-07-12T00:15:54Z  2023-12-06T14:34:06Z  2023-12-06T14:34:06Z    closed   
5  2024-07-12T00:15:53Z  2023-12-06T14:17:41Z  2023-12-06T14:17:41Z    closed   
6  2024-07-12T00:15:53Z  2023-12-06T14:18:21Z  2023-12-06T14:18:21Z    closed   
7  2024-07-12T01:30:39Z  2023-12-06T19:46:39Z  2023-12-06T19:46:38Z    closed   
8  2024-07-12T00:15:52Z  2023-12-06T12:27:43Z  2023-12-06T12:27:43Z    closed   
9  2024-07-12T00:15:52Z  2023-12-08T14:04:26Z  2023-12-08T14:04:26Z    closed   

   project_id  
0        49.0  
1        49.0  
2        49.0  
3        49.0  
4        50.0  
5        50.0  
6        50.0  
7        50.0  
8        50.0  
9        50.0

In [90]:
c.shape

(569448, 8)

In [ ]:
# Check if a column contains NaN
has_nan = c['created_at'].isna().any()

print(has_nan)  # Output: True

False


In [111]:
df=c

# Convert date columns to datetime format
df[['created_at', 'updated_at', 'closed_at', 'merged_at']] = df[['created_at', 'updated_at', 'closed_at', 'merged_at']].apply(pd.to_datetime)

# Assign 'Merged' if 'merged_at' is not NaN, otherwise 'Closed'
df['pr_status'] = df['merged_at'].notna().map({True: 'Merged', False: 'Closed'})

# Make pr_status where both 'closed_at' and 'merged_at' are NaN as 'Open'
df.loc[df[['closed_at', 'merged_at']].isna().all(axis=1), 'pr_status'] = 'Open'

c=df

/tmp/ipykernel_261583/2938148627.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[['created_at', 'updated_at', 'closed_at', 'merged_at']] = df[['created_at', 'updated_at', 'closed_at', 'merged_at']].apply(pd.to_datetime)
/tmp/ipykernel_261583/2938148627.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['pr_status'] = df['merged_at'].notna().map({True: 'Merged', False: 'Closed'})


In [99]:
c.head(10)

project_name              dev_name                created_at  \
0       abdera               dargooo 2020-11-26 21:21:33+00:00   
1       abdera              YYTVicky 2020-03-01 00:37:56+00:00   
2       abdera  nicolasyanncouturier 2013-10-18 13:32:37+00:00   
3       abdera  nicolasyanncouturier 2013-10-14 12:23:03+00:00   
4     accumulo              dlmarion 2023-12-06 14:20:37+00:00   
5     accumulo              dlmarion 2023-12-06 14:02:03+00:00   
6     accumulo              dlmarion 2023-12-06 13:02:38+00:00   
7     accumulo              ctubbsii 2023-12-05 23:12:06+00:00   
8     accumulo              dlmarion 2023-12-05 21:54:27+00:00   
9     accumulo              dlmarion 2023-12-05 20:03:13+00:00   

                 updated_at                 closed_at  \
0 2023-05-21 04:37:38+00:00 2023-05-21 04:37:38+00:00   
1 2020-04-30 02:01:45+00:00                       NaT   
2 2014-07-16 01:44:57+00:00                       NaT   
3 2014-07-13 19:17:32+00:00                       NaT   
4 2024-07-12 00:15:54+00:00 2023-12-06 14:34:06+00:00   
5 2024-07-12 00:15:53+00:00 2023-12-06 14:17:41+00:00   
6 2024-07-12 00:15:53+00:00 2023-12-06 14:18:21+00:00   
7 2024-07-12 01:30:39+00:00 2023-12-06 19:46:39+00:00   
8 2024-07-12 00:15:52+00:00 2023-12-06 12:27:43+00:00   
9 2024-07-12 00:15:52+00:00 2023-12-08 14:04:26+00:00   

                  merged_at pr_status  project_id  
0                       NaT    Closed        49.0  
1                       NaT      Open        49.0  
2                       NaT      Open        49.0  
3                       NaT      Open        49.0  
4 2023-12-06 14:34:06+00:00    Merged        50.0  
5 2023-12-06 14:17:41+00:00    Merged        50.0  
6 2023-12-06 14:18:21+00:00    Merged        50.0  
7 2023-12-06 19:46:38+00:00    Merged        50.0  
8 2023-12-06 12:27:43+00:00    Merged        50.0  
9 2023-12-08 14:04:26+00:00    Merged        50.0

In [ ]:
c.shape

(569448, 8)

**Merging the tables**

In [121]:
e = c.merge(a, on='project_id', how='left')
# Convert 'd' into a dictionary for faster lookup
dev_alias_mapping = dict(zip(d['dev_name'], d['alias_id']))

# Manually search for each dev_name in 'e' within the 'd' dictionary
e['alias_id'] = e['dev_name'].apply(lambda x: dev_alias_mapping.get(x, np.nan))

In [122]:
e.columns

Index(['project_name', 'dev_name', 'created_at', 'updated_at', 'closed_at',
       'merged_at', 'pr_status', 'project_id', 'project_status', 'alias_id'],
      dtype='object')

In [123]:
e.shape

(569448, 10)

In [124]:
# Check if a column contains NaN
has_nan = e['alias_id'].isna().any()

print(has_nan)  # Output: True

True


In [125]:
# Convert b into a lookup dictionary
b_lookup = {(row.project_id, row.alias_id): row.Core_Developer for row in b.itertuples(index=False)}

# Function to determine Core_Developer status
def determine_core_dev(row):
    if pd.isna(row.alias_id):
        return 'No'
    return b_lookup.get((row.project_id, row.alias_id), 'No')

# Apply the function to assign Core_Developer column in e
e['Core_Developer'] = e.apply(determine_core_dev, axis=1)

In [126]:
e.shape

(569448, 11)

In [127]:
e.head(10)

project_name              dev_name                created_at  \
0       abdera               dargooo 2020-11-26 21:21:33+00:00   
1       abdera              YYTVicky 2020-03-01 00:37:56+00:00   
2       abdera  nicolasyanncouturier 2013-10-18 13:32:37+00:00   
3       abdera  nicolasyanncouturier 2013-10-14 12:23:03+00:00   
4     accumulo              dlmarion 2023-12-06 14:20:37+00:00   
5     accumulo              dlmarion 2023-12-06 14:02:03+00:00   
6     accumulo              dlmarion 2023-12-06 13:02:38+00:00   
7     accumulo              ctubbsii 2023-12-05 23:12:06+00:00   
8     accumulo              dlmarion 2023-12-05 21:54:27+00:00   
9     accumulo              dlmarion 2023-12-05 20:03:13+00:00   

                 updated_at                 closed_at  \
0 2023-05-21 04:37:38+00:00 2023-05-21 04:37:38+00:00   
1 2020-04-30 02:01:45+00:00                       NaT   
2 2014-07-16 01:44:57+00:00                       NaT   
3 2014-07-13 19:17:32+00:00                       NaT   
4 2024-07-12 00:15:54+00:00 2023-12-06 14:34:06+00:00   
5 2024-07-12 00:15:53+00:00 2023-12-06 14:17:41+00:00   
6 2024-07-12 00:15:53+00:00 2023-12-06 14:18:21+00:00   
7 2024-07-12 01:30:39+00:00 2023-12-06 19:46:39+00:00   
8 2024-07-12 00:15:52+00:00 2023-12-06 12:27:43+00:00   
9 2024-07-12 00:15:52+00:00 2023-12-08 14:04:26+00:00   

                  merged_at pr_status  project_id  project_status  alias_id  \
0                       NaT    Closed        49.0               1       NaN   
1                       NaT      Open        49.0               1       NaN   
2                       NaT      Open        49.0               1       NaN   
3                       NaT      Open        49.0               1       NaN   
4 2023-12-06 14:34:06+00:00    Merged        50.0               1     340.0   
5 2023-12-06 14:17:41+00:00    Merged        50.0               1     340.0   
6 2023-12-06 14:18:21+00:00    Merged        50.0               1     340.0   
7 2023-12-06 19:46:38+00:00    Merged        50.0               1   16961.0   
8 2023-12-06 12:27:43+00:00    Merged        50.0               1     340.0   
9 2023-12-08 14:04:26+00:00    Merged        50.0               1     340.0   

  Core_Developer  
0             No  
1             No  
2             No  
3             No  
4             No  
5             No  
6             No  
7             No  
8             No  
9             No

In [128]:
# Determining the total Unique values in a column
unique_values = e['Core_Developer'].unique()
print(unique_values)
print(unique_values.size)

['No' 'Yes']
2


In [129]:
# Counting unique values
e['project_status'].value_counts()

project_status
1    374500
0    191219
2      3729
Name: count, dtype: int64

In [130]:
# Save the DataFrame to a CSV file in the current directory
e.to_csv('pr_core_dev.csv', index=False)